e2eAIOK/DeNas/thirdparty/utils.py

```diff
@@ -0,0 +1,132 @@
+    cand_tuple = dict()
+    for search_key in search_space:
+        cand_tuple[search_key] = random.choice(search_space[search_key])
+
+    cand_tuple["hidden_size"] = int(cand_tuple["hidden_size"]/cand_tuple["num_attention_heads"]) * cand_tuple["num_attention_heads"]
+
+    return json.dumps(cand_tuple)
+
+def hf_is_legal(cand, vis_dict, params):
+    if cand not in vis_dict:
+        vis_dict[cand] = {}
+    info = vis_dict[cand]
+    if 'visited' in info:
+        return False
+    cand = json.loads(cand)
+    candidate_net = SuperHFModel.set_sample_config(params.pretrained_model, **cand)
+    n_parameters = sum(param.numel() for param in candidate_net.parameters())
+    info['params'] = n_parameters / 10.**6
+    if info['params'] > params.max_param_limits:
+        return False
+    if info['params'] < params.min_param_limits:
+        return False
+    info['visited'] = True
+    return True
+
+def hf_mutation_random_func(m_prob, s_prob, search_space, top_candidates):
+    cand = random.choice(top_candidates)
+    random_s = random.random()
+    result_cand = json.loads(cand)
+    
+    if random_s < m_prob:
+        # num_heads
+        result_cand['num_attention_heads'] = random.choice(search_space['num_attention_heads'])
+    for search_key in search_space:
+        if search_key == 'num_attention_heads':
+            continue
+        random_s = random.random()
+        if random_s < s_prob:
+            result_cand[search_key] = random.choice(search_space[search_key])
+
+    result_cand["hidden_size"] = int(result_cand["hidden_size"]/result_cand["num_attention_heads"]) * result_cand["num_attention_heads"]
+
+    return json.dumps(result_cand)
+
+def hf_crossover_random_func(top_candidates):
+    p1 = json.loads(random.choice(top_candidates))
+    p2 = json.loads(random.choice(top_candidates))
+    max_iters_tmp = 50
+    while len(p1) != len(p2) and max_iters_tmp > 0:
+        max_iters_tmp -= 1
+        p1 = json.loads(random.choice(top_candidates))
+        p2 = json.loads(random.choice(top_candidates))
+    cand = {}
+    for it1, it2 in zip(p1, p2):
+        select_item = random.choice([p1[it1], p2[it2]])
+        cand[it1] = select_item
+    cand["hidden_size"] = int(cand["hidden_size"]/cand["num_attention_heads"]) * cand["num_attention_heads"]
+    return json.dumps(cand)
+
+def get_hf_latency(model, batch_size, max_seq_length, gpu, infer_cnt):
+    if gpu is None:
+        device = 'cpu'
+    else:
+        device = 'cuda'
+    input_ids = [9333] * max_seq_length
+    input_masks = max_seq_length * [1]
+    input_segments = max_seq_length * [0]
+    input_ids = torch.tensor([input_ids]*batch_size, dtype=torch.long).to(device)
+    input_masks = torch.tensor([input_masks]*batch_size, dtype=torch.long).to(device)
+    input_segments = torch.tensor([input_segments]*batch_size, dtype=torch.long).to(device)
+    batch = {'input_ids': input_ids, 'token_type_ids': input_segments, 'attention_mask':input_masks}
```

Comment:
maybe replace with `input_construtor`?

Reply:
Thanks for the suggestions. I have refined the above issue.